### PR TITLE
7416: Virtual map is unhashed after a reconnect

### PR DESCRIPTION
**Pull request by artemananiev**
Fix summary:

* `hashed` field is removed from `VirtualRootNode`
* Inherited `hash` field isn't used in `VirtualRootNode` to store its hash
* A new `hash` field, of AtomicReference type, is introduced to `VirtualRootNode`. This field is used to store the hash, as well as to check if the root node is hashed or not

Fixes: https://github.com/hashgraph/hedera-services/issues/7416
Signed-off-by: Artem Ananev <artem.ananev@swirldslabs.com>

### CODE REVIEW

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/merkle/VirtualRootNode.java [hunk 1]
```diff
@@ -1212,6 +1213,16 @@ public void setHash(final Hash hash) {
         throw new UnsupportedOperationException("data type is self hashing");
     }
 
+    /**
+     * This class is self-hashing, it doesn't use inherited {@link #setHash} method. Instead,
+     * the hash is set using this private method.
+     *
+     * @param value Hash value to set
+     */
+    private void setHashPrivate(final Hash value) {
```
[the remark this answers is not in the document]
Comment (cody-littley) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/merkle/VirtualRootNode.java:1222]: If you keep this method, please add `@NonNull` annotation.

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/merkle/VirtualRootNode.java [hunk 2]
```diff
@@ -1198,10 +1199,10 @@ public boolean isSelfHashing() {
      */
     @Override
     public Hash getHash() {
-        if (super.getHash() == null) {
+        if (hash.get() == null) {
             pipeline.hashCopy(this);
         }
-        return super.getHash();
+        return hash.get();
```
Comment (cody-littley) [inline platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/merkle/VirtualRootNode.java:1205]: Just double checking. Is `pipeline.hashCopy()` thread safe if called simultaneously by different threads?

**Review by cody-littley — APPROVED** [review]

**Review by OlegMazurov — DISMISSED** [review]

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/merkle/VirtualRootNode.java (hunk 2 above)
Comment (artemananiev) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/merkle/VirtualRootNode.java:1205]: Yes, it is. We had issues in the past caused by calling `pipeline.hashCopy()` in parallel from many threads, this is definitely a possible scenario.

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/merkle/VirtualRootNode.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (artemananiev) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/merkle/VirtualRootNode.java:1222]: In my first version of the fix, `value` was annotated as `@NonNull`, with the corresponding `Objects.requireNonNull()` check. However, in one of the reconnect unit tests the hash is actually null:

```java
    @Test
    @Tags({@Tag("VirtualMerkle"), @Tag("Reconnect"), @Tag("VMAP-003"), @Tag("VMAP-003.7")})
    @DisplayName("Empty teacher and two-leaf learner")
    void emptyTeacherTwoLeafLearner() {
        learnerMap.put(A_KEY, APPLE);
        learnerMap.put(B_KEY, BANANA);
        assertDoesNotThrow(this::reconnect, "Should not throw a Exception");
    }
```

[the remark this answers is not in the document]
Comment (artemananiev) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/merkle/VirtualRootNode.java:1222]: Let me annotate it as `@Nullable` instead.

[the remark this answers is not in the document]
Comment (artemananiev) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/merkle/VirtualRootNode.java:1222]: > Is there any particular reason you need this method? It's private, it has only one line, and this line is trivial.

In runtime, I'm pretty sure JVM will inline this method anyway. In source code, the method is helpful to have as I can attach some JavaDoc to it, annotate params, and so on.

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/merkle/VirtualRootNode.java [hunk 3]
```diff
@@ -822,6 +823,7 @@ public V get(final K key) {
      */
     public void put(final K key, final V value) {
         throwIfImmutable();
+        assert !isHashed() : "Cannot modify already hashed node";
```
[the remark this answers is not in the document]
Comment (artemananiev) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/merkle/VirtualRootNode.java:826]: I wish we had clear policy about when to use assertions vs when to explicitly throw exceptions... In this particular case, the root node may not be already hashed using its `computeHash()` method. Besides that, a hash can be set from `endLearnerReconnect()` or after full leaves rehash, in both cases the root copy is not expected to be mutable. So this assertion is rather a sanity check than something we really expect to happen.

**Review by cody-littley — APPROVED** [review]

**Review by imalygin — APPROVED** [review]

**Review by OlegMazurov — APPROVED** [review]

**Pull request merged**